### PR TITLE
Update AD Graph Cmdlets

### PR DESCRIPTION
**Pull request by shuagarw**
Update AD Graph Cmdlets:
- Upgrade graph sdk to 3.2.0-preview.
- Modify exisitng create app and sp cmdlets.
- Add new app and sp update cmdlets, key credential management cmdlets, and manage user cmdlets
- Catch GraphErrorException in all AD cmdlets and write error with Code and Message from Graph api.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@shuagarw__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (shuagarw). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADAppCredentialCommand.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+using Microsoft.Azure.Graph.RBAC.Models;
+using System;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Creates a new AD application Credential.
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "AzureRmADAppCredential", DefaultParameterSetName = ParameterSet.ApplicationObjectIdWithPassword), OutputType(typeof(PSADCredential))]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADAppCredentialCommand.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADAppCredentialCommand.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                {
+                    // Create object for password credential
+                    var passwordCredential = new PasswordCredential()
+                    {
+                        EndDate = EndDate,
+                        StartDate = StartDate,
+                        KeyId = Guid.NewGuid().ToString(),
+                        Value = Password
+                    };
+
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADAppCredentialCommand.cs]: Protect changes with if (ShouldProcess()) {}, or ConfirmAction()

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADServicePrincipalCommand.cs [hunk 3]
```diff
@@ -22,24 +22,156 @@ namespace Microsoft.Azure.Commands.ActiveDirectory
     /// <summary>
     /// Creates a new service principal.
     /// </summary>
-    [Cmdlet(VerbsCommon.New, "AzureRmADServicePrincipal"), OutputType(typeof(PSADServicePrincipal))]
+    [Cmdlet(VerbsCommon.New, "AzureRmADServicePrincipal", DefaultParameterSetName = ParameterSet.ApplicationWithoutCredential), OutputType(typeof(PSADServicePrincipal))]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADServicePrincipalCommand.cs]: As above, add SuppotsShouldProcess=true, and protect execution with a call to ShouldProcess

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADApplicationCommand.cs [hunk 4]
```diff
@@ -39,19 +39,6 @@ public class NewAzureADApplicationCommand : ActiveDirectoryBaseCmdlet
         public string DisplayName { get; set; }
 
         [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = ParameterSet.ApplicationWithoutCredential,
-            HelpMessage = "The URL to the application’s homepage.")]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADApplicationCommand.cs]: As above, add SuppotsShouldProcess=true to the cmdlet attribute, and protect execution with a call to ShouldProcess

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADSpCredentialCommand.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+using Microsoft.Azure.Graph.RBAC.Models;
+using System;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Creates a new AD servicePrincipal Credential.
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "AzureRmADSpCredential", DefaultParameterSetName = ParameterSet.SpObjectIdWithPassword), OutputType(typeof(PSADCredential))]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADSpCredentialCommand.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADUserCommand.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                },
+                UserPrincipalName = UserPrincipalName
+            };
+
+            if(!string.IsNullOrEmpty(ImmutableId))
+            {
+                userCreateparameters.ImmutableId = ImmutableId;
+            }
+
+            ExecutionBlock(() =>
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADUserCommand.cs]: Protect changes with a call to ShouldProcess(target, action) or ConfirmAction

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/RemoveAzureADAppCredentialCommand.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+using System;
+using System.Management.Automation;
+using ProjectResources = Microsoft.Azure.Commands.Resources.Properties.Resources;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Removes AD application credentials.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "AzureRmADAppCredential", DefaultParameterSetName = ParameterSet.ApplicationObjectIdWithKeyId)]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/RemoveAzureADAppCredentialCommand.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/RemoveAzureADAppCredentialCommand.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                }
+
+                bool deleteAllCredentials = false;
+                if (All.IsPresent)
+                {
+                    deleteAllCredentials = true;
+                }
+
+                if (KeyId != Guid.Empty)
+                {
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/RemoveAzureADAppCredentialCommand.cs]: Guard changes with 
```C#
if(ShouldProcess( target, action)){
/// cmdlet exection
}
```

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/RemoveAzureADSpCredentialCommand.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+using System;
+using System.Management.Automation;
+using ProjectResources = Microsoft.Azure.Commands.Resources.Properties.Resources;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Removes AD SP credentials.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "AzureRmADSpCredential", DefaultParameterSetName = ParameterSet.ObjectIdWithKeyId)]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/RemoveAzureADSpCredentialCommand.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADApplicationCommand.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+using Microsoft.Azure.Commands.Resources.Models.ActiveDirectory;
+using Microsoft.Azure.Graph.RBAC.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Creates a new AD application.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmADApplication"), OutputType(typeof(PSADApplication))]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADApplicationCommand.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADApplicationCommand.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+            HelpMessage = "True if the application is shared with other tenants; otherwise, false.")]
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, ParameterSetName = ParameterSet.ApplicationIdWithUpdateParams,
+            HelpMessage = "True if the application is shared with other tenants; otherwise, false.")]
+        public bool AvailableToOtherTenants { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            ExecutionBlock(() =>
+            {
+                if (!string.IsNullOrEmpty(ApplicationId))
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADApplicationCommand.cs]: Use ```ShouldProcess(string, string)``` or ```ConfirmAction(string, string, action)``` to guard exec

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADServicePrincipalCommand.cs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+using System.Linq;
+using System.Collections.Generic;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Updates an exisitng service principal.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmADServicePrincipal", DefaultParameterSetName = ParameterSet.SpObjectIdWithDisplayName)]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADServicePrincipalCommand.cs]: SupportsShouldProcess=true

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADServicePrincipalCommand.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+                if (sp == null)
+                {
+                    throw new InvalidOperationException("ServicePrincipal does not exist.");
+                }
+
+                // Get AppObjectId
+                string applicationObjectId = ActiveDirectoryClient.GetObjectIdFromApplicationId(sp.ApplicationId.ToString());
+
+                if (!string.IsNullOrEmpty(DisplayName))
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADServicePrincipalCommand.cs]: Guard execution fith ```ShouldProcess(string, string)``` or ```ConfirmAction(string, string, Action)```

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADUserCommand.cs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+using Microsoft.Azure.Commands.Resources.Models.ActiveDirectory;
+using Microsoft.Azure.Graph.RBAC.Models;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Updates an existing AD user.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "AzureRmADUser"), OutputType(typeof(PSADUser))]
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADUserCommand.cs]: ShouldProcess pattern, as with all cmdlets that make changes or have side effects

**Comment by markcowl** [comment]
@shuagarw The issue is that you have removed the quotes from the csv file unintentionally while saving it.  You can import the file into excel, add your columns, and save as a csv.

**Comment by shuagarw** [comment]
@markcowl  Thanks!! I was looking into it and trying to find what I messed up in csv file as it is not able to parse.

**Comment by markcowl** [comment]
@shuagarw if you make the SHouldProcess changes I outlined in review comments, the other static analysis error will go away.  The P0 in SignautreIssues.csv is that your new Remove cmdlets have Force parameters but do not have the SupportsShouldProcess=true property in the cmdlet attribute

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADServicePrincipalCommand.cs (hunk 13 above)
Comment (shuagarw) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADServicePrincipalCommand.cs]: Done.

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADApplicationCommand.cs (hunk 11 above)
Comment (shuagarw) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/SetAzureADApplicationCommand.cs]: Done.

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADUserCommand.cs (hunk 6 above)
Comment (shuagarw) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADUserCommand.cs]: Done.

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADApplicationCommand.cs (hunk 4 above)
Comment (shuagarw) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADApplicationCommand.cs]: Done.

**Comment by shuagarw** [comment]
On-Demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1081/

**Comment by markcowl** [comment]
@shuagarw It looks like it was unable to find the mocks for one of your tests.  This either means  that the file isn't marked as CopyIfNewer or CopyAlways, or the file path is too long and you'll need to shorten the test class or test name

**Comment by cormacpayne** [comment]
On demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1082/

**Comment by shuagarw** [comment]
New on-demand after some small fixes: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1084/

**Comment by shuagarw** [comment]
Powershell -sign: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-sign/559/

**Pull request merged**